output/http/outputhttp.go

```diff
@@ -75,12 +94,28 @@ func InitHandler(
 		TLSClientConfig:       &tls.Config{InsecureSkipVerify: conf.IgnoreSSL},
 	}}
 	conf.control = control
+	conf.queue = make(chan logevent.LogEvent, 5)
+	if conf.MaxQueueSize == 0 || conf.MaxQueueSize < -1 {
+		conf.MaxQueueSize = 1
+	} // we need at least one event in the queue to allow for resume
+	go conf.backgroundtask()
 
 	return &conf, nil
 }
 
-// Output event
+// Output is receiving the event from gogstash handler
 func (t *OutputConfig) Output(ctx context.Context, event logevent.LogEvent) (err error) {
+	// see if output has requested pause and if so just queue the event instead of trying
+	if atomic.LoadUint32(&t.isInPause) == statusPaused {
+		t.queue <- event
```

Comment:
When entered pause mode and the queue is full, `ctx` will not be able to cancel because of blocking here.

Reply:
Is is true that the code will block if the queue gets full. I am not sure the best approach here. I have tried to write the quene handler as fast as possible to minimize this problem.

In case there is congestion this is perhaps the best outcome anyway? The size of the queue channel could be adjusted to the size of the chsize (but that property is not accessible for the output?) or be configured so the user of gogstash can choose for themselves?

Please advice.

Reply:
There is a non-blocking channel pattern in go.

```go
select {
case <- ctx.Done():
	// parent context canceled
case <- channel:
	// process the channel data
default:
	// if the channel is full
}
```

https://stackoverflow.com/questions/52587537/non-blocking-channel-in-go

Reply:
Hi

I don't see any other outcome from this flow than just wait for the event to be put into the queue. So if the code blocks then I think that is ok. The time it takes to queue the event should still be shorter than the normal wait-to-process-an-event.

The output handler (output.go / startOutputs()) blocks further processing until the message is queued or processed. I think that is acceptable.

Please advice.

Reply:
It might be ok in this case because the cooperated functions take care of the blocking issue. (a background go routine consumes the queue)

But I don't think it's a good practice. There is a `ctx` in the parameter, so, at least, you should take care of the parent context canceled case instead of blocking the go routine.

Reply:
I have rewritten the code to handle context now.